Move Http\Client into Http package.

markstory opened the pull request:
Move `Cake\Network\Http\Client` into the Http package. This change just moves the implementation and not the test cases. This shows that the correct class renames are in place. I wanted to move the implementation first to keep this change small and easy to review. I've renamed `FormData\Part` to `FormDataPart` as the extra directories were not necessary.

I also wanted get consensus on the renames before doing more work. After this, I'll be ensuring that the request and response object the client use implement the PSR7 interfaces.

Refs #6960

markstory committed:
Move Http\Client into Http package.

Move Cake\Network\Http\Client into the Http package. This change just
moves the implementation and not the test cases. This shows that the
correct class renames are in place. I wanted to move the implementation
first to keep this change small and easy to review. I've renamed
FormData\Part to FormDataPart as the extra directories were not
necessary.

I also wanted get consensus on the renames before doing more work.

antograssiot commented on src/Http/Client/CookieCollection.php:
```diff
@@ -11,7 +11,9 @@
-namespace Cake\Network\Http;
+namespace Cake\Http\Client;
+
+use Cake\Http\Client\Response;
```
this one isnt necessary

antograssiot commented on src/Http/Client/FormData.php:
```diff
@@ -11,9 +11,9 @@
+namespace Cake\Http\Client;
 
-use Cake\Network\Http\FormData\Part;
+use Cake\Http\Client\FormDataPart;
```
same here

markstory committed:
Remove un-necessary imports.

markstory commented:
@antograssiot Thanks, I've fixed those extra imports up.

lorenzo commented on config/bootstrap.php:
```diff
@@ -29,6 +29,19 @@ class_alias('Cake\Mailer\Email', 'Cake\Network\Email\Email');
+class_alias('Cake\Http\Client\Auth\Basic', 'Cake\Network\Http\Auth\Basic');
+class_alias('Cake\Http\Client\Auth\Digest', 'Cake\Network\Http\Auth\Digest');
+class_alias('Cake\Http\Client\Auth\Oauth', 'Cake\Network\Http\Auth\Oauth');
+class_alias('Cake\Http\Client\FormDataPart', 'Cake\Network\Http\FormData\Part');
```
Do you mind adding `false` as a third argument to all `class_alias` function calls in this file? I think That will prevent the classes from being loaded without a reason, although I'm not 100% sure that is the effect it will have.

I'm just concerned that we are loading a large part of the framework on every request

markstory replied:
Sure thing. I will update the mailer class aliases as well.

markstory replied:
Turning off the autoloading causes a pile of tests to fail.

chinpei215 replied:
Autoloading would be needed in this case. A possible solution would be to keep previous file structures and call `calss_alias()` in the each file, like `Object` in 2.next,
https://github.com/cakephp/cakephp/blob/2.next/lib/Cake/Core/Object.php

or to create `3_2_compat.php`, like a `Chronos`.
https://github.com/cakephp/chronos/blob/master/src/carbon_compat.php

markstory replied:
I'd like to avoid a compat file as the upgrade should be transparent to developers. Adding class stubs might be our best option. I'll give that a shot tonight.

markstory committed:
Don't autoload aliased classes.

Don't load classes that are aliased until they are actually used.

markstory committed:
Revert "Don't autoload aliased classes."

This reverts commit 65c36e533813ee75a8567c5a7fdb1da748c3fb91.

markstory committed:
Move backwards compatible shims.

Instead of eagerly loading the Http client on every request to wire up
class aliases we can get the same effect with stub files that provide
the required aliases.

markstory commented:
@lorenzo @chinpei215 Shim files added now. If we're happy with this approach I can do the same thing for the mailer aliases we added in 3.1.

chinpei215 commented:
Looks good to me.

lorenzo commented:
@markstory doing it for the other files would be nice :)
